File: LeetCode/67. Add Binary/67.cpp
#include <string>

class Solution {
public:
    std::string addBinary(std::string a, std::string b) {
        std::string result;
        int i = a.size() - 1;
        int j = b.size() - 1;
        int carry = 0;

        // Process each bit from the end of both strings
        while (i >= 0 || j >= 0 || carry != 0) {
            int sum = carry;
            if (i >= 0) {
                sum += a[i] - '0'; // Convert char to integer
                i--;
            }
            if (j >= 0) {
                sum += b[j] - '0';
                j--;
            }

            result.push_back((sum % 2) + '0'); // Convert the remainder to char and add to result
            carry = sum / 2; // Update carry
        }

        // The bits are added to the result in reverse order, so we need to reverse the result
        std::reverse(result.begin(), result.end());

        return result;
    }
};

File: LeetCode/67. Add Binary/67.py
class Solution:
    def addBinary(self, a, b):
        # Initialize the indices for both strings
        index_a, index_b = len(a) - 1, len(b) - 1
        carry = 0
        result = []

        # Traverse both strings from right to left
        while index_a >= 0 or index_b >= 0 or carry:
            sum = carry
            if index_a >= 0:
                sum += int(a[index_a])
                index_a -= 1
            if index_b >= 0:
                sum += int(b[index_b])
                index_b -= 1

            # Append current bit to result array
            result.append(sum % 2)
            # Update carry
            carry = sum // 2

        # Reverse the result list to form the final binary string
        return ''.join(str(x) for x in reversed(result))
